### PR TITLE
 brcm63xx: add support for Netcomm NB604N 

### PR DESCRIPTION
**Pull request by jtanx**
Device specifications:
* BCM6328 with BCM43225KMLG wireless NIC
* 8MB flash (S25FL064P)
* 64MB RAM
* 1 USB port, 4 ethernet ports
* 1 RJ11 port for ADSL2

OpenWRT support:
* ADSL unsupported
* Wireless support limited (works with either b43 or brcmsmac,
  but wireless speeds are slow)

Flashing instructions:
* With the device powered off, press and hold the reset button
* Power on the device, keeping the reset button held
* Release the reset button after a while, or when the power led turns blue
* Browse to 192.168.1.1 and upload the OpenWRT CFE image
* The CFE recovery modes listed here all work:
  - https://openwrt.org/docs/techref/bootloader/cfe#bcm63xx_cfe

The stock firmware can be reloaded using the same steps above.

https://openwrt.org/toh/netcomm/nb604n

### CODE REVIEW

**Comment by adrianschmutzler** [comment]
Please add a commit message with device specifications and flashing instructions. For the kernel patches, add a commit message explaining what you change and why.

File: target/linux/brcm63xx/dts/nb604n.dts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+/ {
+	model = "Netcomm NB604N";
+	compatible = "brcm,nb604n", "brcm,bcm6328";
+
+	chosen {
+		bootargs = "rootfstype=squashfs,jffs2 noinitrd console=ttyS0,115200";
+		stdout-path = "serial0:115200n8";
+	};
+
+	gpio-keys-polled {
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/dts/nb604n.dts]: Use "keys"

File: target/linux/brcm63xx/dts/nb604n.dts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+		wps {
+			label = "wps";
+			gpios = <&pinctrl 24 1>;
+			linux,code = <KEY_WPS_BUTTON>;
+			debounce-interval = <60>;
+		};
+	};
+
+	gpio-leds {
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/dts/nb604n.dts]: Use "leds"

File: target/linux/brcm63xx/dts/nb604n.dts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		compatible = "gpio-leds";
+
+		usb {
+			label = "NB604N::usb";
+			gpios = <&pinctrl 1 1>;
+		};
+		inet_red {
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/dts/nb604n.dts]: Add one empty line between blocks (also below).

File: target/linux/brcm63xx/dts/nb604n.dts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+			linux,code = <KEY_WPS_BUTTON>;
+			debounce-interval = <60>;
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		usb {
+			label = "NB604N::usb";
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/dts/nb604n.dts]: Please add the color between the two colons as for all the other devices.

File: target/linux/brcm63xx/dts/nb604n.dts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+		};
+		power_red {
+			label = "NB604N::red:power";
+			gpios = <&pinctrl 5 1>;
+		};
+		wps {
+			label = "NB604N::wps";
+			gpios = <&pinctrl 10 1>;
+		};
+		inet {
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/dts/nb604n.dts]: Use "internet" and add the color for all LEDs that are not unique (like you did for power_red).

File: target/linux/brcm63xx/image/bcm63xx.mk [hunk 6]
```diff
@@ -857,6 +857,20 @@ define Device/livebox
 endef
 TARGET_DEVICES += livebox
 
+### Netcomm ###
+define Device/NB604N
+  $(Device/bcm63xx)
+  IMAGES += sysupgrade.bin
+  DEVICE_TITLE := Netcomm NB604N
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/image/bcm63xx.mk]: Use DEVICE_VENDOR/DEVICE_MODEL instead.

File: target/linux/brcm63xx/image/bcm63xx.mk [hunk 7]
```diff
@@ -857,6 +857,20 @@ define Device/livebox
...
+  DEVICE_DTS := nb604n
+  CFE_BOARD_ID := 96328ang
+  CFE_CHIP_ID := 6328
+  FLASH_MB := 8
+  DEVICE_PACKAGES := \
+    $(BRCMSMAC_PACKAGES) $(USB2_PACKAGES)
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/image/bcm63xx.mk]: This will match into one line.

File: target/linux/brcm63xx/dts/nb604n.dts (hunk 4 above)
Comment (jtanx) [reply target/linux/brcm63xx/dts/nb604n.dts]: Clarifying, do you mean to just explicitly specify the color on all of the leds?

File: target/linux/brcm63xx/image/bcm63xx.mk (hunk 7 above)
Comment (jtanx) [reply target/linux/brcm63xx/image/bcm63xx.mk]: Could you please clarify what you mean by this?

File: target/linux/brcm63xx/dts/nb604n.dts (hunk 4 above)
Comment (adrianschmutzler) [reply target/linux/brcm63xx/dts/nb604n.dts]: For the label, yes. Just as it's done for all the other devices (isn't it?):
NB604N:green:usb or NB604N:red:usb or whatever is true.

File: target/linux/brcm63xx/image/bcm63xx.mk (hunk 7 above)
Comment (adrianschmutzler) [reply target/linux/brcm63xx/image/bcm63xx.mk]: DEVICE_PACKAGES := $(BRCMSMAC_PACKAGES) $(USB2_PACKAGES)

**Comment by adrianschmutzler** [comment]
I've just pushed a tidy-up/reorganization patch for brcm63xx target.
For you, this would mean that you have to update device name, DTS name and the board name in board-files. I will comment inline to make this more clear...

File: target/linux/brcm63xx/image/bcm63xx.mk [hunk 8]
```diff
@@ -857,6 +857,20 @@ define Device/livebox
 endef
 TARGET_DEVICES += livebox
 
+### Netcomm ###
+define Device/NB604N
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/image/bcm63xx.mk]: Device/netcomm_nb604n

File: target/linux/brcm63xx/image/bcm63xx.mk [hunk 9]
```diff
@@ -857,6 +857,20 @@ define Device/livebox
...
+  $(Device/bcm63xx)
+  IMAGES += sysupgrade.bin
+  DEVICE_VENDOR := Netcomm
+  DEVICE_MODEL := NB604N
+  DEVICE_DTS := nb604n
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/image/bcm63xx.mk]: Remove this line, and rename DTS file like soc-vendor-model.dts:
bcm6328-netcomm-nb604.dts

File: target/linux/brcm63xx/image/bcm63xx.mk [hunk 10]
```diff
@@ -857,6 +857,20 @@ define Device/livebox
...
+  CFE_BOARD_ID := 96328ang
+  CFE_CHIP_ID := 6328
+  FLASH_MB := 8
+  DEVICE_PACKAGES := $(B43_PACKAGES) $(USB2_PACKAGES)
+endef
+TARGET_DEVICES += NB604N
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/image/bcm63xx.mk]: same rename as above.

File: target/linux/brcm63xx/base-files/etc/board.d/01_leds [hunk 11]
```diff
@@ -74,6 +74,9 @@ livebox1)
 	ucidef_set_led_netdev "wan" "WAN" "Livebox1:red:adsl" "eth1"
 	ucidef_set_led_netdev "wlan0" "WIFI" "Livebox1:red:wifi" "wlan0"
 	;;
+nb604n)
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/base-files/etc/board.d/01_leds]: netcomm,nb604n)

File: target/linux/brcm63xx/base-files/etc/board.d/02_network [hunk 12]
```diff
@@ -89,6 +89,7 @@ fast2504n |\
 fast2704v2 |\
 hg622 |\
 hg655b |\
+nb604n |\
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/base-files/etc/board.d/02_network]: netcomm,nb604n |\

File: target/linux/brcm63xx/base-files/lib/brcm63xx.sh [hunk 13]
```diff
@@ -180,6 +180,9 @@ brcm63xx_dt_detect() {
 	"Inventel Livebox 1")
 		board_name="livebox1"
 		;;
+	"Netcomm NB604N")
+		board_name="nb604n"
+		;;
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/base-files/lib/brcm63xx.sh]: This file is gone.

File: target/linux/brcm63xx/dts/nb604n.dts [hunk 14]
```diff
@@ -0,0 +1,130 @@
+/dts-v1/;
+
+#include "bcm6328.dtsi"
+
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "Netcomm NB604N";
+	compatible = "brcm,nb604n", "brcm,bcm6328";
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/dts/nb604n.dts]: netcomm,nb604n for the compatible, rename the file as stated below.

File: target/linux/brcm63xx/dts/nb604n.dts [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+	};
+};
+
+&pinctrl {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usb_port1_device>;
+};
+
+&uart0 {
+	status = "ok";
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/dts/nb604n.dts]: IIRC, okay is preferred to ok...

File: target/linux/brcm63xx/patches-4.19/599-board_NB604N.patch [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+ 	&board_R5010UNV2,
++	&board_NB604N,
+ #endif
+ #ifdef CONFIG_BCM63XX_CPU_6338
+ 	&board_96338gw,
+@@ -2829,6 +2873,7 @@ static struct of_device_id const bcm963x
+ 	{ .compatible = "comtrend,ar-5381u", .data = &board_AR5381u, },
+ 	{ .compatible = "comtrend,ar-5387un", .data = &board_AR5387un, },
+ 	{ .compatible = "d-link,dsl-274xb-f", .data = &board_dsl_274xb_f1, },
++	{ .compatible = "netcomm,nb604n", .data = &board_NB604N, },
```
Comment (adrianschmutzler) [inline target/linux/brcm63xx/patches-4.19/599-board_NB604N.patch]: Interesting. This was not matching your compatible before, so it wasn't used at all.

**Comment by Noltari** [comment]
@jtanx in order to merge this PR I would like to know if you are going to send the s25sl064p patch upstream.
It seems that this flash is already used on other OpenWrt devices and I don't want to break them. Is it really needed or is it optional?
Are you going to upstream it?
Thanks!

**Comment by forgeRiddleBoy** [comment]
@jtanx  How stupid are you send an incomplete work and dont respond back for reviews. There are many students(Like me) willing to take this up and would have done it better than your useless stupid work. Dont be a kangaroo to jump here and there and do half work everywhere. idk why these guys choose computer science with no knowledge of it and just want authorities. Go learn some stuffs @jtanx  and dont be oversmart. Thankyou sm @Noltari.

**Comment by adrianschmutzler** [comment]
@forgeRiddleBoy Your comment has been deleted as it clearly violates rule #12 (https://openwrt.org/rules) and does not provide anything useful at all.
You are invited to continue this work when you've learned how to behave.

**Comment by jtanx** [comment]
@forgeRiddleBoy I don't see you contributing anything so kindly gtfo.

As you've probably guessed, circumstances change and I'm no longer in a position to update this any further. 

> It seems that this flash is already used on other OpenWrt devices and I don't want to break them. Is it really needed or is it optional?

iirc it was needed, otherwise nothing gets written back to flash (no settings persist and you can only update via CFE)

> Are you going to upstream it?

No.

Feel free to reopen if anyone actually wants to pick this up.

**Comment by forgeRiddleBoy** [comment]
@jtanx yes you wont be in a position because you're busy linking @ctrlcctrlv & @monkeyiq 's asses to run @fontforge. Go give them a blowjob otherwise they'll kick you out from fontforge. lol truth have been spoken. gtfo? you gtfo motherfucker @jtanx.

If you cannot update this you should have closed this long before Australian kangaroo @jtanx.

@adrianschmutzler  the above comments also violates, I pardon for my words but this motherfucker @jtanx needs this atmost.

**Comment by adrianschmutzler** [comment]
@forgeRiddleBoy We are not interested in trolls here, so please keep away unless you have to contribute something, which will be very welcome if it's provided properly.

**Pull request closed without merging**